### ISSUE TITLE
LDAP: Domain Users are not being loaded with User Filter

### ISSUE THREAD

**Issue by sbelov1** [owncloud/core#13533]
Environment: 
OC v7.0.4 on apache
CentOS 6.5 
php 5.4.32
MySQL
AD LDAP
Description: Customer is attempting to select just a Domain Users group from the User Filter, however only 2 (as oppose to 200) are shown. As a workaround we have to either use a different group or allow all the groups to login with OC.

Here's the log excerpt:

```
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:20+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:20+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:21+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:31+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:42+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:59+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:25:03+00:00"}
{"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:25:04+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/group_ldap.php#477","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/manager.php#100","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#209","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#250","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#136","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/user\/manager.php#140","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/user\/manager.php#142","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"ldap_explode_dn() expects parameter 1 to be string, array given at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#142","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#144","level":3,"time":"2015-01-20T17:31:15+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/manager.php#100","level":3,"time":"2015-01-20T17:31:17+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#209","level":3,"time":"2015-01-20T17:31:17+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-01-20T17:31:17+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#250","level":3,"time":"2015-01-20T17:31:17+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#136","level":3,"time":"2015-01-20T17:31:17+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#137","level":3,"time":"2015-01-20T17:31:17+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/manager.php#100","level":3,"time":"2015-01-20T17:31:21+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#209","level":3,"time":"2015-01-20T17:31:21+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-01-20T17:31:21+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#250","level":3,"time":"2015-01-20T17:31:21+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#136","level":3,"time":"2015-01-20T17:31:21+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#137","level":3,"time":"2015-01-20T17:31:21+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/manager.php#100","level":3,"time":"2015-01-20T17:31:22+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#209","level":3,"time":"2015-01-20T17:31:22+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-01-20T17:31:22+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#250","level":3,"time":"2015-01-20T17:31:22+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#136","level":3,"time":"2015-01-20T17:31:22+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#137","level":3,"time":"2015-01-20T17:31:22+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/group_ldap.php#561","level":3,"time":"2015-01-20T17:32:11+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/group_ldap.php#561","level":3,"time":"2015-01-20T17:32:12+00:00"}
```

@gig13

**Comment by karlitschek** [comment]
@blizzz

**Comment by sbelov1** [comment]
Updating the post with an exact php version:

PHP 5.4.34 (cli) (built: Oct 16 2014 10:19:38)
Copyright (c) 1997-2014 The PHP Group
Zend Engine v2.4.0, Copyright (c) 1998-2014 Zend Technologies
    with Zend Guard Loader v3.3, Copyright (c) 1998-2013, by Zend Technologies

**Comment by blizzz** [comment]
@sbelov1 what are the reproduction steps? What is the LDAP config? 

**Hint**: https://raw.githubusercontent.com/owncloud/core/master/issue_template.md

**Comment by sbelov1** [comment]
@blizzz I believe its the same user we were working with here: https://github.com/owncloud/core/issues/12118

Reproduction steps are as follows:
1. Select Domain Users group only from the Group Filter
2. Select Domain Users group from the User Filter
Result: 2 users are shown
Expected Result: ~200 users

**Comment by sbelov1** [comment]
@blizzz oh yes, and we have requested for LDAP config. I'll update once we get it back from customer.

**Comment by blizzz** [comment]
ok, in the meantime, the  missing users have "Domain Users" assigned as their primary group?

**Comment by sbelov1** [comment]
@blizzz I believe so, yes. Looks like something we've already resolved, didn't we?

**Comment by blizzz** [comment]
i am not aware, just that rework of primary group handling is still on my list (but for other reasons). And unfortunately everything works with my test AD.

**Comment by sbelov1** [comment]
@blizzz would it be easier jumping on a quick call with customer? They've been having this issue for awhile now. It could be something specific to their environment as I'm not having this type of issue with anyone else.

**Comment by blizzz** [comment]
Not really, at least not that alone. I would provide a patch to gather some debug output instead.

**Comment by sbelov1** [comment]
@blizzz sounds good. Should I expect the patch now or do we have to wait for the LDAP config output from customer first?

**Comment by blizzz** [comment]
@sbelov1 I'd like to LDAP have a look on the config first.

**Comment by sbelov1** [comment]
@blizzz please check your email for the output.

**Comment by blizzz** [comment]
@sbelov1: now i remember… the issue here is that the user filter restricts login to a "standard" group membership, but does not take primary groups into account. We sought to fix this with https://github.com/owncloud/core/pull/12233

**Comment by blizzz** [comment]
I follow up there, stay tuned.

**Comment by blizzz** [comment]
Oh, and for a manual workaround: currently there is no group limitation in user or login filter. To make this work with primary groups do as follow:

Change User Filter from

```
(&(|(objectclass=organizationalPerson)(objectclass=person)(objectclass=top)(objectclass=user)))
```

To

```
(&(|(objectclass=organizationalPerson)(objectclass=person)(objectclass=top)(objectclass=user))(|(memberof=CN=Domain Users,…)(primaryGroupID=513)))
```

and Login filter from

```
(&(&(|(objectclass=organizationalPerson)(objectclass=person)(objectclass=top)(objectclass=user)))(|(sAMAccountName=%uid))) 
```

To

```
(&((&(|(objectclass=organizationalPerson)(objectclass=person)(objectclass=top)(objectclass=user))(|(memberof=CN=Domain Users,…)(primaryGroupID=513))))(sAMAccountName=%uid))
```


In both cases you must insert the whole DN of the domain users group in the memberof= pair.

The primaryGroupID should be 513 for Domain Users. Otherwise you can get details of a user's domain group using this script: https://gist.github.com/blizzz/101e17cdc1d399031b50#file-primgrtest-php

Example usage and output:

```
$ sudo -u www-data ./primGrTest.php "cn=tara gilbert,ou=vampires,dc=madder,dc=owncloud,dc=bzoc" s01
User Primary Group ID is: 513
The DN of the domain is: dc=madder,dc=owncloud,dc=bzoc
The SID of the domain is: S-1-5-21-249921958-728525901-1594176202
The object SID should be: S-1-5-21-249921958-728525901-1594176202-513
VERY GOOD: The group DN is: cn=domain users,cn=users,dc=madder,dc=owncloud,dc=bzoc
```

It should not be necessary in this case though.

**Comment by blizzz** [comment]
OK, i think i have a fix here: https://github.com/owncloud/core/pull/13740
(mind: it is against master)

**Comment by sbelov1** [comment]
@blizzz sounds good. Please let me know the instructions for applying it once its all set.

**Comment by blizzz** [comment]
@sbelov1 here is the corresponding PR for stable7: https://github.com/owncloud/core/pull/13742

You can get a diff here https://github.com/owncloud/core/pull/13742.diff, save it and apply it using patch against 7.0.4 (worked for me):

```
$ patch -p1 < 13742.diff
```

**Comment by sbelov1** [comment]
@blizzz this patch wouldn't affect any changes to the database or anything, right? Just a php code? Just wanted to make sure its safe to run as customer is on a production environment.

**Comment by blizzz** [comment]
@sbelov1 exactly. If you took it from before, please save it again, i did a fix there. About code quality in general: yet it has neither not been tested by someone else and from review side got only few comments on coding style and inline documentation.

**Comment by gig13** [comment]
@sbelov1 @blizzz 
This should not be tested in a production environment. Please test internally first with the lab. Also provide a back out process.

**Comment by sbelov1** [comment]
@blizzz I'll let Allen take it from here. Please let me know if there's I can do..

**Comment by blizzz** [comment]
@gig13 Well, testers are necessery. Maybe a test setup can be created there. I tried it with our lab, but I need two more people who give their go in https://github.com/owncloud/core/pull/13742

**Comment by sbelov1** [comment]
@blizzz can you please a clear instruction as to how to run the patch? I've dropped that .diff inside owncloud folder and ran the #patch -p1 < 13742.diff command but nothing happened.

**Comment by gig13** [comment]
@sbelov1 
A quick patch primer http://jungels.net/articles/diff-patch-ten-minutes.html 

Take a look at the patch file, it has the source code file(s) listed and the lines that are modified and updated. Before you run the patch, it is a good idea to make backups of the files, then do the comparison... so you can see the changes applied.

**Comment by gig13** [comment]
@jnfrmarks 
Would you, could you, are you interested in testing? Who else could test? :)

**Comment by sbelov1** [comment]
@blizzz @jnfrmarks @gig13 Just ran the test and it looks like some files didn't get patched:

[root@owncloudtrial owncloud]# patch -p1 < 13742.diff 
(Stripping trailing CRs from patch.)
patching file apps/user_ldap/group_ldap.php
(Stripping trailing CRs from patch.)
patching file apps/user_ldap/lib/access.php
Hunk #1 FAILED at 1069.
1 out of 1 hunk FAILED -- saving rejects to file apps/user_ldap/lib/access.php.rej
(Stripping trailing CRs from patch.)
patching file apps/user_ldap/lib/connection.php
Hunk #1 FAILED at 30.
1 out of 1 hunk FAILED -- saving rejects to file apps/user_ldap/lib/connection.php.rej
(Stripping trailing CRs from patch.)
patching file apps/user_ldap/lib/wizard.php
Hunk #1 succeeded at 803 (offset -1 lines).
(Stripping trailing CRs from patch.)
patching file apps/user_ldap/tests/group_ldap.php
patch unexpectedly ends in middle of line
Hunk #1 succeeded at 77 with fuzz 1.

**Comment by blizzz** [comment]
did you re-apply it on the already patched file from https://github.com/owncloud/core/issues/13533#issuecomment-72099771 ?

**Comment by sbelov1** [comment]
@blizzz the patch was never applied before since I didn't have 'patch' module installed. This was the first time I actually ran it.

**Comment by sbelov1** [comment]
@blizzz any other ideas?
@gig did you get any luck running it or is it just me?

**Comment by cdamken** [comment]
My idea could be:

User Filter:
```
(|(memberof=CN=Domain Users,…)(primaryGroupID=513))
```
Login filter:
```
(|(memberof=CN=Domain Users,…)(primaryGroupID=513)(sAMAccountName=%uid))
```

**Comment by blizzz** [comment]
Mh, okay, applying patch failed here, too, most likely because of additional changes after my test. I will prepare one for 7.0.4

**Comment by blizzz** [comment]
@sbelov1 please try with (file also available on S3 under Shared/owncloud/support/Patches/13742.704.patch) :

```diff
diff --git a/apps/user_ldap/group_ldap.php b/apps/user_ldap/group_ldap.php
index e8d268d..aba87e5 100644
--- a/apps/user_ldap/group_ldap.php
+++ b/apps/user_ldap/group_ldap.php
@@ -249,32 +249,75 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
 	}
 
 	/**
-	 * returns a list of users that have the given group as primary group
+	 * returns a filter for a "users in primary group" search or count operation
 	 *
 	 * @param string $groupDN
-	 * @param $limit
-	 * @param int $offset
-	 * @return string[]
+	 * @param string $search
+	 * @return string
+	 * @throws \Exception
 	 */
-	public function getUsersInPrimaryGroup($groupDN, $limit = -1, $offset = 0) {
+	private function prepareFilterForUsersInPrimaryGroup($groupDN, $search = '') {
 		$groupID = $this->getGroupPrimaryGroupID($groupDN);
 		if($groupID === false) {
-			return array();
+			throw new \Exception('Not a valid group');
 		}
 
-		$filter = $this->access->combineFilterWithAnd(array(
-			$this->access->connection->ldapUserFilter,
-			'primaryGroupID=' . $groupID
-		));
+		$filterParts = [];
+		// part for counting users (see countUsers in user backend)
+		// it is consolidated in OC 8. No big changes for OC 7.
+		$filterParts[] = \OCP\Util::mb_str_replace(
+			'%uid', '*', $this->access->connection->ldapLoginFilter, 'UTF-8');
+		if(!empty($search)) {
+			$search = $this->access->escapeFilterPart($search, true);
+			$filterParts[] = $this->access->getFilterPartForUserSearch($search);
+		}
+		$filterParts[] = 'primaryGroupID=' . $groupID;
+
+		$filter = $this->access->combineFilterWithAnd($filterParts);
 
-		$users = $this->access->fetchListOfUsers(
-			$filter,
-			array($this->access->connection->ldapUserDisplayName, 'dn'),
-			$limit,
-			$offset
-		);
+		return $filter;
+	}
+
+	/**
+	 * returns a list of users that have the given group as primary group
+	 *
+	 * @param string $groupDN
+	 * @param string $search
+	 * @param int $limit
+	 * @param int $offset
+	 * @return string[]
+	 */
+	public function getUsersInPrimaryGroup($groupDN, $search = '', $limit = -1, $offset = 0) {
+		try {
+			$filter = $this->prepareFilterForUsersInPrimaryGroup($groupDN, $search);
+			return $this->access->fetchListOfUsers(
+				$filter,
+				array($this->access->connection->ldapUserDisplayName, 'dn'),
+				$limit,
+				$offset
+			);
+		} catch (\Exception $e) {
+			return array();
+		}
+	}
 
-		return $users;
+	/**
+	 * returns the number of users that have the given group as primary group
+	 *
+	 * @param string $groupDN
+	 * @param string $search
+	 * @param int $limit
+	 * @param int $offset
+	 * @return int
+	 */
+	public function countUsersInPrimaryGroup($groupDN, $search = '', $limit = -1, $offset = 0) {
+		try {
+			$filter = $this->prepareFilterForUsersInPrimaryGroup($groupDN, $search);
+			$users = $this->access->countUsers($filter, array('dn'), $limit, $offset);
+			return (int)$users;
+		} catch (\Exception $e) {
+			return 0;
+		}
 	}
 
 	/**
@@ -405,6 +448,7 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
 		if(!$this->groupExists($gid)) {
 			return array();
 		}
+		$search = $this->access->escapeFilterPart($search, true);
 		$cacheKey = 'usersInGroup-'.$gid.'-'.$search.'-'.$limit.'-'.$offset;
 		// check for cache of the exact query
 		$groupUsers = $this->access->connection->getFromCache($cacheKey);
@@ -473,7 +517,7 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
 		$groupUsers = array_slice($groupUsers, $offset, $limit);
 
 		//and get users that have the group as primary
-		$primaryUsers = $this->getUsersInPrimaryGroup($groupDN, $limit, $offset);
+		$primaryUsers = $this->getUsersInPrimaryGroup($groupDN, $search, $limit, $offset);
 		$groupUsers = array_unique(array_merge($groupUsers, $primaryUsers));
 
 		$this->access->connection->writeToCache($cacheKey, $groupUsers);
@@ -512,10 +556,13 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
 		}
 
 		if(empty($search)) {
-			$groupUsers = count($members);
+			$primaryUsers = $this->countUsersInPrimaryGroup($groupDN, '');
+			$groupUsers = count($members) + $primaryUsers;
+
 			$this->access->connection->writeToCache($cacheKey, $groupUsers);
 			return $groupUsers;
 		}
+		$search = $this->access->escapeFilterPart($search, true);
 		$isMemberUid =
 			(strtolower($this->access->connection->ldapGroupMemberAssocAttr)
 			=== 'memberuid');
@@ -557,10 +604,9 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
 		}
 
 		//and get users that have the group as primary
-		$primaryUsers = $this->getUsersInPrimaryGroup($groupDN);
-		$groupUsers = array_unique(array_merge($groupUsers, $primaryUsers));
+		$primaryUsers = $this->countUsersInPrimaryGroup($groupDN, $search);
 
-		return count($groupUsers);
+		return count($groupUsers) + $primaryUsers;
 	}
 
 	/**
@@ -623,6 +669,7 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
 		if(!$this->enabled) {
 			return array();
 		}
+		$search = $this->access->escapeFilterPart($search, true);
 		$pagingSize = $this->access->connection->ldapPagingSize;
 		if ((! $this->access->connection->hasPagedResultSupport)
 		   	|| empty($pagingSize)) {
diff --git a/apps/user_ldap/lib/access.php b/apps/user_ldap/lib/access.php
index 3ff1a99..ed62744 100644
--- a/apps/user_ldap/lib/access.php
+++ b/apps/user_ldap/lib/access.php
@@ -1084,7 +1084,7 @@ class Access extends LDAPUtility implements user\IUserTools {
 	/**
 	* escapes (user provided) parts for LDAP filter
 	* @param string $input, the provided value
-	* @param bool $allowAsterisk wether in * at the beginning should be preserved
+	* @param bool $allowAsterisk whether in * at the beginning should be preserved
 	* @return string the escaped string
 	*/
 	public function escapeFilterPart($input, $allowAsterisk = false) {
diff --git a/apps/user_ldap/lib/connection.php b/apps/user_ldap/lib/connection.php
index 336ea7b..ed3f1d3 100644
--- a/apps/user_ldap/lib/connection.php
+++ b/apps/user_ldap/lib/connection.php
@@ -29,6 +29,9 @@ namespace OCA\user_ldap\lib;
  * @property string ldapUserFilter
  * @property string ldapUserDisplayName
  * @property boolean hasPagedResultSupport
+ * @property string[] ldapBaseUsers
+ * @property string ldapLoginFilter
+ * @property string ldapGroupMemberAssocAttr
 */
 class Connection extends LDAPUtility {
 	private $ldapConnectionRes = null;
diff --git a/apps/user_ldap/lib/wizard.php b/apps/user_ldap/lib/wizard.php
index b08516a..9518aa0 100644
--- a/apps/user_ldap/lib/wizard.php
+++ b/apps/user_ldap/lib/wizard.php
@@ -803,13 +803,23 @@ class Wizard extends LDAPUtility {
 						}
 						$base = $this->configuration->ldapBase[0];
 						foreach($cns as $cn) {
-							$rr = $this->ldap->search($cr, $base, 'cn=' . $cn, array('dn'));
+							$rr = $this->ldap->search($cr, $base, 'cn=' . $cn, array('dn', 'primaryGroupToken'));
 							if(!$this->ldap->isResource($rr)) {
 								continue;
 							}
 							$er = $this->ldap->firstEntry($cr, $rr);
+							$attrs = $this->ldap->getAttributes($cr, $er);
 							$dn = $this->ldap->getDN($cr, $er);
-							$filter .= '(memberof=' . $dn . ')';
+							if(empty($dn)) {
+								continue;
+							}
+							$filterPart = '(memberof=' . $dn . ')';
+							if(isset($attrs['primaryGroupToken'])) {
+								$pgt = $attrs['primaryGroupToken'][0];
+								$primaryFilterPart = '(primaryGroupID=' . $pgt .')';
+								$filterPart = '(|' . $filterPart . $primaryFilterPart . ')';
+							}
+							$filter .= $filterPart;
 						}
 						$filter .= ')';
 					}
diff --git a/apps/user_ldap/tests/group_ldap.php b/apps/user_ldap/tests/group_ldap.php
index d1262e4..7bad37c 100644
--- a/apps/user_ldap/tests/group_ldap.php
+++ b/apps/user_ldap/tests/group_ldap.php
@@ -77,10 +77,15 @@ class Test_Group_Ldap extends \PHPUnit_Framework_TestCase {
 			->method('readAttribute')
 			->will($this->returnValue(array('u11', 'u22', 'u33', 'u34')));
 
+		// for primary groups
+		$access->expects($this->once())
+			->method('countUsers')
+			->will($this->returnValue(2));
+
 		$groupBackend = new GroupLDAP($access);
 		$users = $groupBackend->countUsersInGroup('group');
 
-		$this->assertSame(4, $users);
+		$this->assertSame(6, $users);
 	}
 
 	public function testCountWithSearchString() {
```

**Comment by sbelov1** [comment]
@cdamken thanks Carlos. But we're looking for more of a permanent solution here.

**Comment by blizzz** [comment]
The solution is in the mentioned PRs -.-

**Comment by sbelov1** [comment]
@blizzz hmm... now I did apply the patch (which partially failed) to this VM before. Not sure if it has anything to do with the issues I'm having now..

[root@owncloudtrial owncloud]# patch -p1 < 13742.704.patch
patching file apps/user_ldap/group_ldap.php
Reversed (or previously applied) patch detected!  Assume -R? [n] y
patching file apps/user_ldap/lib/access.php
Hunk #1 FAILED at 1084.
1 out of 1 hunk FAILED -- saving rejects to file apps/user_ldap/lib/access.php.r                  ej
patching file apps/user_ldap/lib/connection.php
patching file apps/user_ldap/lib/wizard.php
Reversed (or previously applied) patch detected!  Assume -R? [n] n
Apply anyway? [n] y
Hunk #1 FAILED at 803.
1 out of 1 hunk FAILED -- saving rejects to file apps/user_ldap/lib/wizard.php.r                  ej
patching file apps/user_ldap/tests/group_ldap.php
Reversed (or previously applied) patch detected!  Assume -R? [n] n
Apply anyway? [n] y
Hunk #1 FAILED at 77.
1 out of 1 hunk FAILED -- saving rejects to

**Comment by blizzz** [comment]
>  Reversed (or previously applied) patch detected! Assume -R? [n] y

o.O it was not a clean 7.0.4, was it?

**Comment by sbelov1** [comment]
correct. I've applied a previous patch you've provided.

**Comment by blizzz** [comment]
well, please go to original 7.0.4 and apply it against this.

**Comment by sbelov1** [comment]
@blizzz Looks to be OK after applying to the fresh build. 

[root@owncloudtrial owncloud]# patch -p1 < 13742.704.patch
patching file apps/user_ldap/group_ldap.php
patching file apps/user_ldap/lib/access.php
patching file apps/user_ldap/lib/connection.php
patching file apps/user_ldap/lib/wizard.php
patching file apps/user_ldap/tests/group_ldap.php

 However I won't be sure if it actually fixes the issue since I wasn't able to reproduce on my end. @gig13 do you want to apply the patch with customer now?

**Comment by sbelov1** [comment]
In my case it shows that all 8 Domain Users have been brought down from AD.

![image](https://cloud.githubusercontent.com/assets/8365936/6062914/6ac870ac-ad21-11e4-9f04-72f0679ecada.png)

**Comment by sbelov1** [comment]
Actually, wait. It shows 6 DU as oppose to 8 shown in the Filter plus the numbers on the Nav panel don't seem to make sense.

![image](https://cloud.githubusercontent.com/assets/8365936/6062976/d81fea7c-ad21-11e4-95de-e5f55f9440fc.png)

**Comment by blizzz** [comment]
@sbelov1 could you run this script: https://gist.github.com/blizzz/736339dc4938a920e087#file-getusergroupmemberships-php from within the owncloud root please and show the output?

usage:

```
$ php -f getUserGroupMemberships.php john.smith "Domain _Users"
```

**Comment by sbelov1** [comment]
@blizzz 
John Smith is member of following groups:
Domain Users

**Comment by blizzz** [comment]
And the "Domain Users has following members:" part?

**Comment by sbelov1** [comment]
Oh, didn't notice it as it was in line with the command prompt:

John Smith is member of following groups:
Domain Users
Invalid GroupID[root@owncloudtrial owncloud]#

**Comment by blizzz** [comment]
@sbelov1 can you copy the whole command you used?

**Comment by sbelov1** [comment]
It was the one you've provided.
# php -f getUserGroupMembership.php john.smith "Domain _Users"

**Comment by blizzz** [comment]
@sbelov1 plus remove the underscore. Sorry for the typo.

**Comment by sbelov1** [comment]
Getting the same thing:

John Smith is member of following groups:
Domain Users
Invalid GroupID

**Comment by blizzz** [comment]
And the "Domain Users has following members:" part?

**Comment by blizzz** [comment]
I meant,  can you copy the whole command you used?

**Comment by sbelov1** [comment]
ok, it looks like it never got to that part of the code after throwing an 'Invalid GroupID' message. So, I've placed that part of the code just above it and here's what I'm getting now:
Domain Users

Domain Users has following members:
(none)

**Comment by blizzz** [comment]
:question: :exclamation: what was the command you used?

**Comment by sbelov1** [comment]
It was the same command:

$php -f getUserGroupMemberships.php john.smith "Domain_Users"

**Comment by blizzz** [comment]
try with:

```
$php -f getUserGroupMemberships.php john.smith "Domain Users"
```

**Comment by sbelov1** [comment]
oops, sorry. Here's the one with removed underscore.

John Smith is member of following groups:
Domain Users
Domain Users has following members:
(none)

**Comment by sbelov1** [comment]
@blizzz any updates?

**Comment by sbelov1** [comment]
@blizzz having the same issue with yet another prospect. They were running an old appliance with 7.0.3 and we've upgraded them to 7.0.4. User groups are not being rendered for most of the users even though respective groups have been selected in both User and Group filters.
![users](https://cloud.githubusercontent.com/assets/8365936/6256986/97e34b52-b78a-11e4-8a12-546c7aa42c5e.JPG)
  Sending you an LDAP config over email. This is a regular setup: CentOS 6.5, MySQL, Apache, php 54.
 On the side note, since we're using an old appliance, are there any additional libraries that need to be changed outside of ownCloud? Maybe some of them are simply missing while we've only updated an ownCloud code and nothing else.

**Comment by craigpg** [comment]
#### here's what I saw
I logged into the ownCloud instance that @sbelov1 reported had missing groups despite those groups being selected in the wizard and observed the following behavior:
1. "group_xyz" (not the real group name) was not appearing on the users page even though it was believed be selected in the group filters
2. in the group filter tab, there were too many groups selected to see all of them clearly in the "only from those groups" selector:
![image](https://cloud.githubusercontent.com/assets/160667/6359771/612fdb44-bc43-11e4-9362-ee4160ab5b19.png)
3. I clicked on "edit raw filter instead" and "group_xyz" was NOT in the raw filter, which of course would prevent it from showing up on the users page.
4. I clicked on "edit raw filter instead" again to toggle back to using the menu selectors.
5. I clicked on the "only from those groups" selector to get the scrolling list of checkboxes to see what was actually selected.  It turns out that "group_xyz" _was_ checked.  I made _no_ changes to the settings, I simply scrolled through the list of checkboxes.
5. I clicked on "edit raw filter instead" again and "group_xyz" _was_ in the raw filter.
6. I then went to the users page and "group_xyz" was there, as expected.

#### conclusion
There appears to be a case where the menu selectors get out of sync with the generated raw filter.  I believe in all cases, the raw filter is the source of truth while the UI wizardry, which makes things easier for non-LDAP admin types, needs to be double-checked.  It's likely there some subtle timing issues that can result in the wizard and raw filter becoming out of sync.

Anyway, the above steps should be validated by someone else but here's my takeaway.
1) Double-check raw filter to make sure it matches what you believe you selected in the wizard.  I know it has a lot of parens and nested expressions, but it's pretty self-explanatory.
2) Toggling the raw filter and reactivating / scrolling the "only from those groups" selector might get things back in sync.  Try it and then take another look at the raw filter.

Finally, I know where making wizard improvements, especially w.r.t. group selection, in 8.1

**Comment by craigpg** [comment]
@blizzz, and ideas on what can cause the errors(?) we're seeing in the logs:
> {"app":"PHP","message":"ldap_search(): Partial search results returned: Sizelimit exceeded at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/ldap.php#252","level":3,"time":"2015-01-20T17:24:42+00:00"}

Is this something we can safely ignore or does it suggest that something is misconfigured and that we're potentially not getting complete results?

**Comment by blizzz** [comment]
@craigpg When the group is selected, it should be also in the group filter, no doubt. However, this is a different issue and does not relate to the original problem. Requires a new bug report. Ideally with reproduction steps. 

If I understand correctly, with the correct filter this works as expected, right?

**Comment by sbelov1** [comment]
@blizz sorry, not sure exactly what you mean by a "proper filter". Yes this is separate issue and I thought I've had a bug created for this.

**Comment by sbelov1** [comment]
Here it is https://github.com/owncloud/core/issues/12118

**Comment by LukasReschke** [comment]
@blizz is still not working on ownCloud and you probably meant to refer to @blizzz (3 z) instead :smile: 

As this notifies the wrong blizz all the time this is something unfortunate. Well, unless the wrong blizz wants to work on the ownCloud project as well? :see_no_evil:

**Comment by sbelov1** [comment]
Haha, sorry @blizzz didn't mean to insult you, mate! ✋

**Comment by blizz** [comment]
It would be better for me if you abandoned the thread and started a new one when this happens.


Sent from my Verizon Wireless 4G LTE smartphone

<div>-------- Original message --------</div><div>From: Lukas Reschke <notifications@github.com> </div><div>Date:02/24/2015  5:43 PM  (GMT-07:00) </div><div>To: owncloud/core <core@noreply.github.com> </div><div>Cc: blizz <blizzardville@yahoo.com> </div><div>Subject: Re: [core] LDAP: Domain Users are not being loaded with User Filter
  (#13533) </div><div>
</div>@blizz is still not working on ownCloud and you probably meant to refer to @blizzz (3 z) instead 

As this notifies the wrong blizz all the time this is something unfortunate. Well, unless the wrong blizz wants to work on the ownCloud project as well?

**Comment by sbelov1** [comment]
or we can just update our comments ;)

**Comment by blizzz** [comment]
@sbelov1 this won't have effect.

**Comment by blizzz** [comment]
> @blizzz sorry, not sure exactly what you mean by a "proper filter". Yes this is separate issue and I thought I've had a bug created for this.

The filter appeared correct, but was in fact wrong as your mentioned group was missing in it. As i stated in the  email and @craigpg in his comment, too.

**Comment by sbelov1** [comment]
btw @craigpg, thanks for the assessment. That's exactly the behavior I'm experiencing.

**Comment by blizzz** [comment]
@sbelov1 i.e. it works and we can  close the issue?

**Comment by sbelov1** [comment]
@blizzz no, this is just an assessment of the issue, Arthur. You tell me when we can close it. We're on standby awaiting for the fix as check boxes are inconstant with the Raw Filters.

**Comment by blizzz** [comment]
@sbelov1 this is a separate, unrelated issue. Within this ticket, we will not fix all possible bugs in LDAP,  user management and bootstrapping. You problem was that  group associations  did not work  when groups were primary groups in AD, and that user filter creation did not include primary groups. This was to be solved.

**Comment by sbelov1** [comment]
actually, @craigpg @blizzz I've just checked this again. And here's what I'm getting:

1. Selected the Raw Filter format in the Group filter
2. Noticed that Cloud_Instructors group WAS present in the filter
3. Users Page shows only 2 users being members of the Cloud_Instructors group where in reality, there should be at least 4. Which number, by the way is shown in the nav. panel on the left!
![image](https://cloud.githubusercontent.com/assets/8365936/6374453/02523152-bcdf-11e4-98ef-689cad885e5d.png)
 So Raw Filter DOES seem to be in sync in this case, while users are still not shown as members of their respective groups within oC. Furthermore, if I were to present a CIFS share to Could_Instructors group, it would only be shown to 2 people, not 4. We've tested that with the prospect before.

**Comment by sbelov1** [comment]
@blizzz I agree, this is related to https://github.com/owncloud/core/issues/12118 as mentioned before. We can continue the thread there if you'd like.

**Comment by blizzz** [comment]
> Noticed that Cloud_Instructors group WAS present in the filter

Of  course. @craigpg made the wizard recompile it…

> Users Page shows only 2 users being members of the Cloud_Instructors group where in reality, there should be at least 4. Which number, by the way is shown in the nav. panel on the left! 

Correct. The LDAP  backend however detects the group correctly, as the server returns them all:

```
{"data":[{"name":"David","displayname":"David Pershteyn","groups":["admin","Cloud_EMS","Cloud_FOD","Cloud_FSA","Cloud_instructors"],"subadmin":[],"quota":"default","storageLocation":"\/mnt\/ocdata\/David","lastLogin":"1423596257"},{"name":"JMcBride","displayname":"James McBride","groups":["Cloud Hazmat"],"subadmin":[],"quota":"default","storageLocation":"\/mnt\/ocdata\/JMcBride","lastLogin":"1423773004"},{"name":"JPriest","displayname":"John Priest","groups":["Cloud Hazmat"],"subadmin":[],"quota":"default","storageLocation":"\/mnt\/ocdata\/JPriest","lastLogin":"1424375207"},{"name":"RNoonan","displayname":"Robert Noonan","groups":["Cloud_instructors"],"subadmin":[],"quota":"default","storageLocation":"\/mnt\/ocdata\/RNoonan","lastLogin":"1423837487"}],"status":"success"}
```

I.e. this is a  bug in the users page and actually that was addressed imho, maybe not backported? I will look for the corresponding PR.

Btw/FYI, I installed openldap-clients on the server so i could use ldapsearch.

**Comment by sbelov1** [comment]
> Correct. The LDAP backend however detects the group correctly, as the server returns them all:
Hmm... if it does detect the groups correctly, why then the CIFS share only available to fewer folks than it should? It appears that only the people who are shown as members of the group (on the Users Page) are actually the only ones who you can share the files/folders via group membership.

**Comment by blizzz** [comment]
Sorry, I misread, they are indeed missing for two users. :see_no_evil:

**Comment by blizzz** [comment]
test scripts retrieves them properly … 

```
sudo -u apache php -f ./group-memberships.php JMcBride Cloud_instructors
James McBride is member of following groups:
Cloud Hazmat
Cloud_instructors

Cloud_instructors has following members:
David Pershteyn (David)
James McBride (JMcBride)
John Priest (JPriest)
Robert Noonan (RNoonan)
```

will do more debugging later

**Comment by blizzz** [comment]
I found the cause of the trouble. A paged search will not be initialized if a page size is not specified, as it happens when we look up groups. However, if you run paged searches, you must initiliaze it, because otherwise LDAP server will refer to the old one. In this case, we have had a prior paged search where we only looked for one result, this is why only a subset (well one) of the groups is returned.

Although this requires certain circumstances it is weird that we did not see this happen earlier. 

I'll work on a fix tomorrow.

**Comment by sbelov1** [comment]
@blizzz @MorrisJobke sounds good. Please keep us posted guys. We have some prospects/customers waiting on this fix as I've stated before.

**Comment by blizzz** [comment]
> Although this requires certain circumstances it is weird that we did not see this happen earlier. 

One reason is that at least this case is not affected by OpenLDAP (well, usually, it can with a different, not so usual setting) but AD is.

**Comment by blizzz** [comment]
I've wrote a test that resembles the same conditions. AD, having a paged result with limit of 1 before and, fetching the user groups. This is supposed too fail to of course. However, it does not, but succeeds. *sigh*

@sbelov1 i'd like to try it on your machine, but it times out right now.

**Comment by sbelov1** [comment]
@blizzz we no longer have that environment available, unfortunately. So it would be great if you could test another LDAP instance. Let me know if you couldn't reproduce it in your environment, I can get an access to mine, where I can actually see the same thing happening.

**Comment by sbelov1** [comment]
@blizzz any progress? Btw, the prospect says he wasn't aware that connection was lost even though I've sent an email. I'll keep you posted once we have a connection back. In the meantime, feel free using my instance for testing.

**Comment by sbelov1** [comment]
Actually, @blizzz, please try prospect's instance again. It should be available now. I was able to access it.

**Comment by blizzz** [comment]
@sbelov1 today i could access that instance and my test script for this issue works here as expected (i.e. fails), unlike on my setup. That's good in the way that I am able to isolate this stuff in a reproducible test. The bad thing is, that it is bound to some other factors, possibly external. Needs more detective work… Do you know which AD version he has running?

**Comment by blizzz** [comment]
@sbelov1 on your instance,  i see users with one group in total only…?

**Comment by sbelov1** [comment]
@blizzz One group is shown in the Users page

![image](https://cloud.githubusercontent.com/assets/8365936/6485108/38b656f4-c24f-11e4-9598-7dce3ef90615.png)

even though two groups were selected in the wizard

![image](https://cloud.githubusercontent.com/assets/8365936/6485094/2154e232-c24f-11e4-9e02-cb8c9976c2af.png)

**Comment by sbelov1** [comment]
@blizzz Arthur, since we have prospect's instance, I would like to encourage you working on that first since people are waiting. Hopefully, the patch would work for the other instance as well.

**Comment by sbelov1** [comment]
@blizzz prospect's using Win 2012 R2 and my AD is set up on 2008 R2 Datacenter.

**Comment by blizzz** [comment]
Okay. I tried it again against a lab instance with 2008 and my local instance of 2012, and still could not reproduce it. Also compared configurations, but no suspicious difference… I guess I will just write the fix, but having not a reliable test available to avoid regressions in future is tiresome. Maybe I'll work around it with a different, inderirect assertion.

**Comment by sbelov1** [comment]
@blizzz sounds good. Please keep me posted.

**Comment by blizzz** [comment]
@sbelov1 my fix (OC 7 version) is here: https://github.com/owncloud/core/pull/14707

I tested it already on that remote server and my test script succeed, also group assignments in the UI look right. Can you confirm (and see whether everything else still does fine)?

**Comment by sbelov1** [comment]
@blizz looks good to me. I'll have the prospect do some testing and will let you know. Is it something that that would need to be backported into 7.0.5?

**Comment by blizzz** [comment]
@sbelov1  IIRC 7.0.5 is frozen, would be 7.0.6 then. Correct me if i am wrong, @DeepDiver1975.

**Comment by sbelov1** [comment]
@blizzz btw, the patch worked on my internal instance as well! So I'm very hopeful this gets backported in at some point.

**Comment by blizzz** [comment]
@sbelov1 very good, thank you for testing and confirming. I am glad this story here is coming to an end finally :)

**Comment by sbelov1** [comment]
Yup. let me just verify this with the prospect before we can close the GH.

**Comment by blizzz** [comment]
@sbelov1 how does it look like, can we close it?

**Comment by blizzz** [comment]
I close it. Feel free to reopen if the issue unexpectedly persists.

**Comment by gig13** [comment]
@blizzz 
This is still an issue. I upgraded  to OC7.0.5 and tested the patch access.php file.

The LDAP still does not display the users in the Users screen, and the User Login still only finds 2 users in the Domain Users group, instead of the 269 that actually exist.

When viewing the Users page there was a brief refresh of the screen and for a second it shows the 269 Domain Users, and then after completion it only shows 467 Users and only 2 Domain Users. Users can login, but the administrator is unable to make changes easily due to the UI issues.

Latest owncloud.log and Report configuration are available on s3:
https://s3.owncloud.com/owncloud/index.php/apps/files?dir=%2FShared%2Fowncloud%2Fsupport%2Fgithub-issues%2Fcore%2F13533 

If this can not be resolved, a screen share will be needed.

**Comment by blizzz** [comment]
I won't be able  to look at it before Thu.

Well, it was working before on the test scenario and on the remote server. I would kindly ask you to test again, restart the web server after upgrade, apply the patch, wait for Cache timeout (default 10min) and revise the configuration.

**Comment by gig13** [comment]
@blizzz 
What do you mean restart after the upgrade? I am not planning on upgrading them again, they are at OC705 and using the access.php for that version.

**Comment by gig13** [comment]
@blizzz 
ping!

**Comment by gig13** [comment]
ping!

**Comment by blizz** [comment]
Please close this thread and start one without including me in it.
ThanksBrett
 


     On Monday, April 6, 2015 7:29 AM, gig13 <notifications@github.com> wrote:
   

 ping!—

**Comment by MorrisJobke** [comment]
@blizz You can unsubscribe in the right panel. It's a simple button.

**Comment by LukasReschke** [comment]
This might also help for the whole repository. Though untested: ![2015-04-07_10-01-01](https://cloud.githubusercontent.com/assets/878997/7019138/181b6ee2-dd0d-11e4-8f36-e4256ff19c27.png)

**Comment by gig13** [comment]
@blizzz 
Have you had a chance to look at the logs?

**Comment by blizzz** [comment]
@gig13 

> What do you mean restart after the upgrade? 

Well, you said you did upgrade to 7.0.5 (https://github.com/owncloud/core/issues/13533#issuecomment-87062555).


Did you do 

>  I would kindly ask you to test again, restart the web server after upgrade, apply the patch, wait for Cache timeout (default 10min) and revise the configuration.

as suggested in https://github.com/owncloud/core/issues/13533#issuecomment-87623862 ?

**Comment by blizzz** [comment]
Your log is full of repeated messages like:

```
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#181","level":3,"time":"2015-03-27T14:21:46+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/user_ldap.php#182","level":3,"time":"2015-03-27T14:21:46+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/manager.php#100","level":3,"time":"2015-03-27T14:21:46+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#209","level":3,"time":"2015-03-27T14:21:46+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-03-27T14:21:46+00:00"}
{"app":"PHP","message":"Illegal offset type in isset or empty at \/var\/www\/html\/owncloud\/lib\/private\/user\/database.php#250","level":3,"time":"2015-03-27T14:21:46+00:00"}
```

And a few of

```
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/group_ldap.php#561","level":3,"time":"2015-03-27T14:15:56+00:00"}
```

Plus also several of

```
{"app":"wnd","message":"Exception: Couldn't get xattr for smb:\/\/10.1.1.27\/ONE\/PATH\/TO\/A\/FILE: unknown error (2)","level":4,"time":"2015-03-27T14:20:57+00:00"}
```

Already the beginning indicates that invalid user name data is passed, not even a string but an array. 

Adding to the answers to the questions in the prior post, i'd like to ask you how you did the upgrade and how you test it.

**Comment by gig13** [comment]
@blizzz 
Yes, I did the suggestion from  #13533 (comment). I reconfigured the LDAP several times and waited more than ten minutes for the refresh.

I performed an upgrade from 7.0.4 to 7.0.5 in a manual manner as per our documented instructions.  Then applied the patch and reconfigured. And we tested it by logging in as the admin user and as an LDAP user and tested uploading, shares, etc.

**Comment by blizzz** [comment]
Do you have the patch you applied?

**Comment by gig13** [comment]
@blizzz 
Its the access.php in https://s3.owncloud.com/owncloud/index.php/apps/files?dir=%2FShared%2Fowncloud%2Fsupport%2Fgithub-issues%2Fcore%2F13533

**Comment by blizzz** [comment]
OK, I got the missing piece.

**Comment by gig13** [comment]
@blizzz 
Is there a new access.php file?

**Comment by blizzz** [comment]
@gig13 there is one open pull request that needs testers, but as it seems it fixes only one  aspect (counting): https://github.com/owncloud/core/pull/13742  

The missing part I am working on, not being, but likely won't finish today.

**Comment by blizzz** [comment]
@gig13 please try with https://github.com/owncloud/core/pull/13742, I updated it with the missing bits.

**Comment by gig13** [comment]
@blizzz 
I updated and tested with the patch of 5 files (See 20150414_patch_apps.zip on s3) and the Display count is correct for the Domain Users in the LDAP Wizard and Users page!! Unfortunately, not all of the users show up in the Users page, many can not be searched, and the lazy loading did not appear to work... thus the initial result set of users are the only ones displayed.

We changed the settings a couple times in the LDAP wizard configurations, and also waited at least 10 minutes for changes, but there was the same behavior.

I uploaded the latest config_report_20140415.txt and owncloud.log to s3.

https://s3.owncloud.com/owncloud/index.php/apps/files?dir=%2FShared%2Fowncloud%2Fsupport%2Fgithub-issues%2Fcore%2F13533

**Comment by blizzz** [comment]
> See 20150414_patch_apps.zip on s3

Does not seem to be there. File search does not find it.

> Unfortunately, not all of the users show up in the Users page, 

Which are missing?

> many can not be searched

Might be a side effect.

In the logfile, there are is a couple of new repeating entries, but they appear after the background job. This might lead to caching issues:

{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/apps\/user_ldap\/lib\/jobs.php#89","level":3,"time":"2015-04-15T13:40:37+00:00"}
{"app":"PHP","message":"Array to string conversion at \/var\/www\/html\/owncloud\/3rdparty\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Statement.php#138","level":3,"time":"2015-04-15T13:40:37+00:00"}

I was able to reproduce it once, but only once. It might lead thereto that a limited set of users is cached which could end up in displaying the users not correctly. So far I did not see how/why this happened, I am looking further into it.

**Comment by gig13** [comment]
@blizzz 
The  20150414_patch_apps.zip is now on s3.

**Comment by blizzz** [comment]
@gig13 still was unable to reproduce the background job error. but it does not convince me anyway.

Could apply following patch and retry? 

```diff
diff --git a/apps/user_ldap/group_ldap.php b/apps/user_ldap/group_ldap.php
index 10e8bfd..38bade8 100644
--- a/apps/user_ldap/group_ldap.php
+++ b/apps/user_ldap/group_ldap.php
@@ -514,10 +514,10 @@ class GROUP_LDAP extends BackendUtility implements \OCP\GroupInterface {
            }
        }
 
+       $groupUsers = array_unique(array_merge($groupUsers, $primaryUsers));
        natsort($groupUsers);
        $this->access->connection->writeToCache('usersInGroup-'.$gid.'-'.$search, $groupUsers);
        $groupUsers = array_slice($groupUsers, $offset, $limit);
-       $groupUsers = array_unique(array_merge($groupUsers, $primaryUsers));
 
        $this->access->connection->writeToCache($cacheKey, $groupUsers);
 
```

In case it's a cache issues this might solve it. I have not seen anything else so far.

**Comment by craigpg** [comment]
@gig13, I think @blizzz has provided some questions and next steps.

> Unfortunately, not all of the users show up in the Users page, 

Which are missing?

Also, has the above patch been applied and retried?

**Comment by bboule** [comment]
This is waiting on @gig13 test

**Comment by gig13** [comment]
@craigpg @bboule @blizzz 
I tested oc705 and saw the same issue as the customer -- Domain Users were not being found correctly in the Wizard or in the Users page.
After the latest patch was applied, the Domain Users are correctly found (47 in our test instance) and the Users page displays them correctly when only filtering on that group. I tried with IE and Firefox and both browsers displayed the Users page fine. 

I will coordinate a time to install with the customer.

**Comment by sbelov1** [comment]
@blizzz I have another prospect experiencing same/similar issue. A user selects Domain Users group -> only 2 users are shown under User Filter (should be much more) -> 2 users are shown under Users Page -> Only those two users can login. Does it sound like the same issue? If it does, can we please update the patch to 8.0.2 because that's the version we're using. Unless we ported it into 8.0.3 in which case, I'll just update their build.

**Comment by sbelov1** [comment]
@blizzz ?

**Comment by bboule** [comment]
My understanding is that a fix given to @gig13 solved this issue for a customer running 7.0.5 can we confirm the same fix will apply in this case @sbelov1 I think you referred to 8.0.3.  Have we applied to an 8.0.3 instance internally to test?

**Comment by jnfrmarks** [comment]
@bboule 

I can get you an 8.0.3 for testing 13533 if that helps

**Comment by MorrisJobke** [comment]
> My understanding is that a fix given to @gig13 solved this issue for a customer running 7.0.5 can we confirm the same fix will apply in this case @sbelov1 I think you referred to 8.0.3. Have we applied to an 8.0.3 instance internally to test?

Can we stop apply random patches to production instances. If the patch is verified to fix an issue, we should get it into the next patch release and provide that for a production environment.

**Comment by karlitschek** [comment]
I agree with @MorrisJobke We have to handle such problems  this more coordinated and professionally

**Comment by gig13** [comment]
@bboule @MorrisJobke @sbelov1 @karlitschek 
The patch for OC7.0.5 has been successfully applied with the customer environment and is closed from my side. 
The customer had several failed patches attempted (and reverted) over a 4 month period and this was critical for them to properly administer and thus made it very difficult to open up to additional users.

**Comment by sbelov1** [comment]
@bboule We've tested it against 8.0.2 and it did not work. That is why I wanted to confirm with @blizzz on whether this was ported into 8.0.3 before upgrading them.

**Comment by craigpg** [comment]
@blizzz, has this patch made it into a release yet?

**Comment by blizzz** [comment]
If you referring to https://github.com/owncloud/core/issues/13533#issuecomment-94026394, this still waits for a test result from @gig13. 

If you refer to the patch mentioned in https://github.com/owncloud/core/issues/13533#issuecomment-93977253 it probably refers to #13742. Which is a backport of #13740 which made it to 8.0.2. However, the patch in https://github.com/owncloud/core/issues/13533#issuecomment-94026394 would need to be applied to OC 8 as well. I will do a PR out of it.

**Comment by gig13** [comment]
@blizzz 
The patch worked for OC7.0.5 internally -- and was applied at the customer as stated previously.

I just tested OC8.0.3 (no patching) and it does not work. @sbelov1  is testing against OC8, I was testing against OC7.

**Comment by sbelov1** [comment]
> Which is a backport of #13740 which made it to 8.0.2

This is odd. The prospect having this issue was running 8.0.2. However, I've just applied this patch in my internal lab and it did fix the issue. Are we sure it made it to 8.0.2?

**Comment by blizzz** [comment]
> This is odd. The prospect having this issue was running 8.0.2. However, I've just applied this patch in my internal lab and it did fix the issue. Are we sure it made it to 8.0.2?

> From the notes on #13740 it looks like its in 8.1 milestone. Please clarify.

Indeed, it was not backported to stable8.

**Comment by blizzz** [comment]
Patch from https://github.com/owncloud/core/issues/13533#issuecomment-94026394 is added to #15606 (eventually needs to be backported to OC 8, too).

For OC 7 it is added to https://github.com/owncloud/core/pull/13742 which incorporates the backport of #15606 already, since the issues are connected and it was open for so long.

**Comment by MorrisJobke** [comment]
Backport is in #16456
